### PR TITLE
refactor: :recycle: Extract filter to api call

### LINKED ISSUE #7: Make filter into an API call
Use api to filter and use debouncer for search input to improve performance

### PR DESCRIPTION
**Pull request by dannyyys**
Implement #7

### CODE REVIEW

**Comment by OkkarMin** [comment]
Hey @dannyyys it seems like the site is not working, no matter what type of option I use to filter, it will show me the 'not found' image, please do check

👉 https://where-to-mug-git-refactor-extractfiltertoapi-okkar.vercel.app is the preview URL

Looks like there is an error with the API call from the frontend

Screenshot 2021-07-02 at 12 03 52

Might have something to do with the `server` config variable

File: config.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+const dev = process.env.NODE_ENV !== "production";
+
+export const server = dev ? "http://localhost:3000" : "https://wheretomug.ml";
```
Comment (OkkarMin) [inline config.js]:
```ts
export const server = dev ? "http://localhost:3000" : "https://wheretomug.ml";
```

I suspect it has got to do with this server variable

https://vercel.com/docs/environment-variables#system-environment-variables

Comment (dannyyys) [reply config.js]: The preview is still not working. I think it is cause of the preview server url "https://where-to-mug-git-refactor-extractfiltertoapi-okkar.vercel.app". 
I have set the server url in production to be the normal url "https://wheretomug.ml".
Screenshot_20210702-143233_Chrome

Comment (OkkarMin) [reply config.js]: Could you check it you are able to utilize the built-in env vars?

https://vercel.com/docs/environment-variables#system-environment-variables

Comment (dannyyys) [reply config.js]: Updated and committed change.

Comment (OkkarMin) [reply config.js]: The idea is to dynamically get the deployment URL if it's not local dev

File: src/components/FreeRoomsCardList.tsx [hunk 2]
```diff
@@ -1,6 +1,8 @@
 import { FC } from "react";
 
-import { SimpleGrid } from "@chakra-ui/react";
+import { useState, useEffect } from "react";
```
Comment (OkkarMin) [inline src/components/FreeRoomsCardList.tsx]:
```suggestion
import { FC, useState, useEffect } from "react";
```

File: src/components/FreeRoomsCardList.tsx [hunk 3]
```diff
@@ -11,26 +13,44 @@ import room_ids from "../../data/room_ids.json";
...
-      timeSlot == "All" ? true : room_occupancy[stringDay][room][timeSlot];
-
-    return hasRoomName && hasAvailableSlot;
-  });
+  currentDay: string;
+}> = ({ searchText, timeSlot, currentDay }) => {
+  const [fliteredRooms, setFliteredRooms] = useState([]);
+  const [loading, setLoading] = useState(false);
+
+  const URLsearchText = encodeURIComponent(searchText);
```
Comment (OkkarMin) [inline src/components/FreeRoomsCardList.tsx]: try to put this line just before fetch occurs and

rename `URLsearchText` to `encodedSearchText`

File: src/components/FreeRoomsCardList.tsx [hunk 4]
```diff
@@ -11,26 +13,44 @@ import room_ids from "../../data/room_ids.json";
...
+
+  useEffect(() => {
+    const api = async () => {
```
Comment (OkkarMin) [inline src/components/FreeRoomsCardList.tsx]: rename `api` to be more descriptive

File: src/components/FreeRoomsCardList.tsx [hunk 5]
```diff
@@ -11,26 +13,44 @@ import room_ids from "../../data/room_ids.json";
...
+    };
+    api();
+  }, [searchText, timeSlot, currentDay]);
 
   const haveRooms = fliteredRooms.length > 0;
 
-  return haveRooms ? (
+  return loading ? (
+    <Box mt="20" width="full" align="center">
+      {<Spinner size="xl" />}
```
Comment (OkkarMin) [inline src/components/FreeRoomsCardList.tsx]: Why is the `{ }` there for?

File: src/components/Main.tsx [hunk 6]
```diff
@@ -22,17 +23,22 @@ const numberDay = new Date().getDay();
...
+  const handleDebounce = useCallback(
+    debounce((e) => setSearchText(e.target.value), 1000),
+    []
+  );
+
+  const handleSearchTextChange = (e: ChangeEvent<any>) => handleDebounce(e);
   const handleTimeSlotChange = (e: ChangeEvent<any>) =>
     setTimeSlot(e.target.value);
-  const handleDayChange = (e: ChangeEvent<any>) =>
+  const handleDayChange = (e: ChangeEvent<any>) => {
```
Comment (OkkarMin) [inline src/components/Main.tsx]: remove `{ }`

File: src/pages/api/index.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+  const fliteredRooms = rooms.filter((room) => {
+    const hasRoomName = room.includes(searchText.toUpperCase());
+    const hasAvailableSlot =
+      timeSlot == "ALL" ? true : room_occupancy[currentDay][room][timeSlot];
+
+    return hasRoomName && hasAvailableSlot;
+  });
+
+  res.status(200).json(fliteredRooms);
```
Comment (OkkarMin) [inline src/pages/api/index.js]:
```suggestion
  return res.status(200).json(fliteredRooms);
```

**Review by OkkarMin — CHANGES_REQUESTED** [review]
Please update

**Review by OkkarMin — APPROVED** [review]
LGTM!

**Pull request merged**